zookeeper/core/field.py

```diff
@@ -0,0 +1,230 @@
+import inspect
+from typing import Callable, Generic, Type, TypeVar, Union
+
+from zookeeper.core import utils
+from zookeeper.core.partial_component import PartialComponent
+
+
+# A sentinel class/object for missing default values.
+class _MISSING:
+    def __repr__(self):
+        return f"<missing>"
+
+
+_missing = _MISSING()
+
+
+# Type-variables to parameterise fields.
+_ComponentType = TypeVar("_ComponentType")
+_FieldType = TypeVar("_FieldType")
+
+
+class Field(Generic[_ComponentType, _FieldType]):
+    """
+    A configurable field for Zookeeper components. Fields must be typed, may
+    take default values, and are configurable through the CLI.
+
+    This class is not appropriate for nesting child sub-components; for this,
```

Comment:
Could you explain the difference between `Field` and `ComponentField`?

Reply:
1) The form of arguments and the argument validation is different, to support things like `PartialComponent`.
2) The behaviour of `get_default` is slightly different in that in the latter case it will instantiate a component instance for you.
3) I think it's less confusing for users to make it very obvious which fields will have the component inheritence stuff applied.